# Plotting 2D SFS on Wahab

## Gazza minuta

### Make SFS with EASYSFS

```sh
salloc
module load container_env python3
cd easySFS
crun.python3 -p ~/.conda/envs/easySFS ./easySFS.py -i ~/PIRE/gazza_minuta/gmi.all.A.nohighhet.Ham.vcf -p Gmi_popfile.txt --preview -a #sampling ALL SNPs

#AHam downsampling results
(2, 756) (3, 1134) (4, 1421) (5, 1664) (6, 1877) (7, 2069) (8, 2245) (9, 2407)
(10, 2558) (11, 2700) (12, 2833) (13, 2959) (14, 3079) (15, 3193) (16, 3302) 
(17, 3405) (18, 3505) (19, 3599) (20, 3691) (21, 3775) (22, 3860) (23, 3936)
(24, 4015) (25, 4048) (26, 4121) (27, 4079) (28, 4146) (29, 3908) (30, 3966)
(31, 3417) (32, 3465) (33, 2616) (34, 2651) (35, 1563) (36, 1583)

#CBat downsampling results
(2, 763) (3, 1144) (4, 1440) (5, 1693) (6, 1918) (7, 2123) (8, 2312) (9, 2488)
(10, 2653) (11, 2809) (12, 2956) (13, 3096) (14, 3230) (15, 3358) (16, 3480)
(17, 3597) (18, 3710) (19, 3818) (20, 3923) (21, 4023) (22, 4120) (23, 4214)
(24, 4305) (25, 4393) (26, 4478) (27, 4560) (28, 4640) (29, 4717) (30, 4793)
(31, 4866) (32, 4937) (33, 5006) (34, 5073) (35, 5138) (36, 5202) (37, 5264)
(38, 5324) (39, 5383) (40, 5440) (41, 5496) (42, 5550) (43, 5603) (44, 5655)
(45, 5706) (46, 5755) (47, 5800) (48, 5847) (49, 5893) (50, 5938) (51, 5981)
(52, 6024) (53, 6065) (54, 6106) (55, 6147) (56, 6186) (57, 6225) (58, 6262)
(59, 6297) (60, 6333) (61, 6366) (62, 6400) (63, 6428) (64, 6462) (65, 6483)
(66, 6515) (67, 6526) (68, 6557) (69, 6551) (70, 6580) (71, 6536) (72, 6564)
(73, 6474) (74, 6501) (75, 6283) (76, 6308) (77, 5961) (78, 5984) (79, 5410)
(80, 5430) (81, 4357) (82, 4373) (83, 2722) (84, 2732)

#AHam: 28 "individuals" (really 2x individuals) retains most sites
#CBat: 72 retains most sites

crun.python3 -p ~/.conda/envs/easySFS ./easySFS.py -i ~/PIRE/gazza_minuta/gmi.all.A.nohighhet.Ham.vcf -p Gmi_popfile.txt --proj 28,72 -o Gmi_output_28_72 -a -f
```

### Plot with DADI

```sh
cd Gmi_output_28_72/dadi
crun.python3 -p ~/.conda/envs/dadi python
import matplotlib.pyplot as pyplot
import dadi
data = dadi.Spectrum.from_file("AHam-CBat.sfs")
data.sample_sizes #make sure read in right, should match what downsample projections were

fig = pyplot.figure(219033)
fig.clear()
dadi.Plotting.plot_single_2d_sfs(data, vmin=0.01) #vmin means won't plot allele freqs <0.01 (bc log scale, would look funky)
fig.savefig("Gmi_2D_SFS_28_72.png")
```

## Equulites laterofenestra

## Make SFS with EASYSFS

```sh
salloc
module load container_env python3
cd easySFS
crun.python3 -p ~/.conda/envs/easySFS ./easySFS.py -i ~/PIRE/leiognathus_leuciscus/lle.all.ela.nohighhet.vcf -p Lle_popfile.txt --preview -a #sampling ALL SNPs

#AHam
(2, 6262) (3, 9393) (4, 11662) (5, 13500) (6, 15069) (7, 16451) (8, 17693) 
(9, 18824) (10, 19864) (11, 20619) (12, 21511) (13, 21782) (14, 22553) (15, 22539)
(16, 23210) (17, 22637) (18, 23219) (19, 21186) (20, 21665)

#CNas
(2, 5880) (3, 8821) (4, 10933) (5, 12631) (6, 14072) (7, 15335) (8, 16465)
(9, 17491) (10, 18433) (11, 19305) (12, 20118) (13, 20879) (14, 21595) (15, 22271)
(16, 22911) (17, 23520) (18, 24099) (19, 24652) (20, 25181) (21, 25688) (22, 26174)
(23, 26641) (24, 27090) (25, 27523) (26, 27940) (27, 28343) (28, 28732) (29, 29108)
(30, 29472) (31, 29825) (32, 30167) (33, 30498) (34, 30820) (35, 31132) (36, 31436)
(37, 31731) (38, 32018) (39, 32297) (40, 32569) (41, 32833) (42, 33091) (43, 33342)
(44, 33587) (45, 33826) (46, 34059) (47, 34287) (48, 34509) (49, 34726) (50, 34938)
(51, 35145) (52, 35347) (53, 35545) (54, 35738) (55, 35927) (56, 36113) (57, 36294)
(58, 36471) (59, 36644) (60, 36814) (61, 36980) (62, 37143) (63, 37303) (64, 37459)
(65, 37612) (66, 37762) (67, 37910) (68, 38054) (69, 38195) (70, 38334) (71, 38470)
(72, 38603) (73, 38734) (74, 38863) (75, 38989) (76, 39112) (77, 39234) (78, 39353)
(79, 39470) (80, 39585) (81, 39697) (82, 39808) (83, 39917) (84, 40024) (85, 40128)
(86, 40232) (87, 40333) (88, 40432) (89, 40530) (90, 40626) (91, 40720) (92, 40813)
(93, 40902) (94, 40992) (95, 41077) (96, 41163) (97, 41243) (98, 41327) (99, 41405)
(100, 41486) (101, 41559) (102, 41638) (103, 41709) (104, 41784) (105, 41855)
(106, 41928) (107, 41993) (108, 42063) (109, 42123) (110, 42192) (111, 42249)
(112, 42315) (113, 42368) (114, 42433) (115, 42486) (116, 42548) (117, 42598)
(118, 42658) (119, 42707) (120, 42766) (121, 42816) (122, 42872) (123, 42918)
(124, 42973) (125, 43002) (126, 43055) (127, 43087) (128, 43138) (129, 43177)
(130, 43227) (131, 43253) (132, 43301) (133, 43318) (134, 43365) (135, 43376)
(136, 43422) (137, 43434) (138, 43478) (139, 43496) (140, 43539) (141, 43544)
(142, 43585) (143, 43574) (144, 43614) (145, 43599) (146, 43637) (147, 43628)
(148, 43666) (149, 43645) (150, 43681) (151, 43668) (152, 43703) (153, 43677)
(154, 43711) (155, 43674) (156, 43707) (157, 43681) (158, 43713) (159, 43674)
(160, 43705) (161, 43656) (162, 43686) (163, 43618) (164, 43647) (165, 43584)
(166, 43613) (167, 43517) (168, 43545) (169, 43425) (170, 43452) (171, 43355)
(172, 43380) (173, 43248) (174, 43273) (175, 43105) (176, 43129) (177, 42953)
(178, 42976) (179, 42690) (180, 42712) (181, 42344) (182, 42365) (183, 41591)
(184, 41612) (185, 39383) (186, 39402)

#AHam: 18 "individuals" (really 2x individuals) retains most sites
#CBat: 160 retains most sites

crun.python3 -p ~/.conda/envs/easySFS ./easySFS.py -i ~/PIRE/leiognathus_leuciscus/lle.all.ela.nohighhet.vcf -p Lle_popfile.txt --proj 18,160 -o Lle_output_18_160 -a -f
```

### Plot with DADI

```sh
cd Lle_output_18_160/dadi
crun.python3 -p ~/.conda/envs/dadi python
import matplotlib.pyplot as pyplot
import dadi
data = dadi.Spectrum.from_file("AHam-CNas.sfs")
data.sample_sizes #make sure read in right, should match what downsample projections were

fig = pyplot.figure(219033)
fig.clear()
dadi.Plotting.plot_single_2d_sfs(data, vmin=0.01) #vmin means won't plot allele freqs <0.01 (bc log scale, would look funky)
fig.savefig("Lle_2D_SFS_18_160.png")
```
