Read depth plots:
 * **Ela_depth_dist.png:** Histogram of read depth for *Equulites laterofenestra*. Created with `sequencing_stats.R`.
 * **Ela_perchet_depth.png:** Plot of the proportion of all genotypes that were heterozygous at any given read depth for *Equulites laterofenestra*. Created with `sequencing_stats.R`.
 * **Gmi_depth_dist.png:** Histogram of read depth for *Gazza minuta*. Created with `sequencing_stats.R`.
 * **Gmi_perchet_depth.png:** Plot of the proportion of all genotypes that were heterozygous at any given read depth for *Gazza minuta*. Created with `sequencing_stats.R`.
 * **Read_ratio_Ela.png:** Density plot of read depth ratio for *Equulites laterofenestra*. Read depth ratio is calculated on a per locus basis, and equals the average read count for all homozygous genotypes divided by the averag eread count for all heterozygous genotypes. Created with `sequencing_stats.R`.
 * **Read_ratio_Gmi.png:** Density plot of read depth ratio for *Gazza minuta*. Read depth ratio is calculated on a per locus basis, and equals the average read count for all homozygous genotypes divided by the averag eread count for all heterozygous genotypes. Created with `sequencing_stats.R`.

All other plots:
 * **Ela_mean_AB_plot.png:** Density plot of average allele balance across individuals per locus for *Equulites laterofenestra*. Dashed lines indicate cut-offs for allele balance filtering. Created with `sequencing_stats.R`.
 * **Gmi_mean_AB_plot.png:** Density plot of average allele balance across individuals per locus for *Gazza minuta*. Dashed lines indicate cut-offs for allele balance filtering. Created with `sequencing_stats.R`.
 * **missing_data_Ela.png:** Heat map of missing genotypes per individual for *Equulites laterofenestra*. Rows represent loci, columns represent individuals. Cells are colored if a genotype was called and empty (white) if the genotype was missing. Darker columns represent historical individuals, lighter columns represent contemporary individuals. Created with `sequencing_stats.R`.
 * **missing_data_Gmi.png:** Heat map of missing genotypes per individual for *Gazza minuta*. Rows represent loci, columns represent individuals. Cells are colored if a genotype was called and empty (white) if the genotype was missing. Darker columns represent historical individuals, lighter columns represent contemporary individuals. Created with `sequencing_stats.R`.
      
